crypto/x509/v3_ncons.c

```diff
@@ -305,6 +305,134 @@ int NAME_CONSTRAINTS_check(X509 *x, NAME_CONSTRAINTS *nc) {
+  if (dnsid == NULL || idlen == NULL) {
+    return X509_V_ERR_UNSPECIFIED;
+  }
+
+  // Don't leave outputs uninitialized
+  *dnsid = NULL;
+  *idlen = 0;
+  
+  // Per RFC 6125, DNS-IDs representing internationalized domain names appear
+  // in certificates in A-label encoded form:
+  //
+  // https://tools.ietf.org/html/rfc6125#section-6.4.2
+  //
+  // The same applies to CNs which are intended to represent DNS names.
+  // However, while in the SAN DNS-IDs are IA5Strings, as CNs they may be
+  // needlessly encoded in 16-bit Unicode.  We perform a conversion to UTF-8
+  // to ensure that we get an ASCII representation of any CNs that are
+  // representable as ASCII, but just not encoded as ASCII.  The UTF-8 form
+  // may contain some non-ASCII octets, and that's fine, such CNs are not
+  // valid legacy DNS names.
+  //
+  // Note, 'int' is the return type of ASN1_STRING_to_UTF8() so that's what
+  // we must use for 'utf8_length'.
+  unsigned char *utf8_value;
+  int utf8_length = ASN1_STRING_to_UTF8(&utf8_value, cn);
+  if (utf8_length < 0) {
+    return X509_V_ERR_OUT_OF_MEM;
+  }
+  
+  // Some certificates have had names that include a *trailing* NUL byte.
+  // Remove these harmless NUL characters. They would otherwise yield false
+  // alarms with the following embedded NUL check.
+  while (utf8_length > 0 && utf8_value[utf8_length - 1] == '\0') {
+    --utf8_length;
+  }
+
+  // Reject *embedded* NULs
+  if (OPENSSL_memchr(utf8_value, 0, utf8_length) != NULL) {
+    OPENSSL_free(utf8_value);
+    return X509_V_ERR_UNSUPPORTED_NAME_SYNTAX;
+  }
+
+  int isdnsname = 0;
+  
+  // XXX: Deviation from strict DNS name syntax, also check names with '_'
+  // Check DNS name syntax, any '-' or '.' must be internal,
+  // and on either side of each '.' we can't have a '-' or '.'.
+  //
+  // If the name has just one label, we don't consider it a DNS name.  This
+  // means that "CN=sometld" cannot be precluded by DNS name constraints, but
+  // that is not a problem.
+  for (int i = 0; i < utf8_length; ++i) {
+    unsigned char c = utf8_value[i];
+
+    if ((c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') ||
+        (c >= '0' && c <= '9') || c == '_') {
+      continue;
+    }
+
+    // Dot and hyphen cannot be first or last.
+    if (i > 0 && i < utf8_length - 1) {
+      if (c == '-') {
+        continue;
+      }
+      
+      // Next to a dot the preceding and following characters must not be
+      // another dot or a hyphen.  Otherwise, record that the name is
+      // plausible, since it has two or more labels.
+      if (c == '.' && utf8_value[i + 1] != '.' && utf8_value[i - 1] != '-' &&
+          utf8_value[i + 1] != '-') {
+        isdnsname = 1;
+        continue;
+      }
+    }
+    isdnsname = 0;
+    break;
+  }
+
+  if (isdnsname) {
+    *dnsid = utf8_value;
+    *idlen = (size_t)utf8_length;
+    return X509_V_OK;
+  }
+  OPENSSL_free(utf8_value);
+  return X509_V_OK;
```

Comment:
NP: For names containing NULs we return `X509_V_ERR_UNSUPPORTED_NAME_SYNTAX` to kick us out of the for loop in the caller. For all other invalid names, we return `X509_V_OK`.
* DNS name are limited to like ~255 characters -- should we also reject names that are too long?
* Besides containing a NUL, are there other *egrecious* issues we might find with a name for which we should return that as well?